protocols/kad/src/behaviour.rs

```diff
@@ -293,9 +293,9 @@ impl<TSubstream> Kademlia<TSubstream> {
     ///
     /// The actual meaning of *providing* the value of a key is not defined, and is specific to
     /// the value whose key is the hash.
-    pub fn add_providing(&mut self, key: PeerId) {
-        self.providing_keys.insert(key.clone().into());
-        let providers = self.values_providers.entry(key.into()).or_insert_with(Default::default);
+    pub fn add_providing(&mut self, key: &Multihash) {
```

Comment:
What about `Multihash` instead of `&Multihash`, so that we remove the `clone`?
```suggestion
    pub fn add_providing(&mut self, key: Multihash) {
```